### PR TITLE
Implementation for batched metrics

### PR DESCRIPTION
**Pull request by bryanlou**
**Description:** 
Batch similar metrics originating from Prometheus or Open Telemetry to send existing log events for CloudWatch more efficiently over AOC EMF exporter

**Testing:** 
Add some unit tests for generated GroupMetrics and processing OpenTelemetry metrics to see if it's processed correctly

**Documentation:**
None

### CODE REVIEW

File: exporter/awsemfexporter/metric_translator.go [hunk 1]
```diff
@@ -140,48 +140,116 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, config *Config) ([]*CWMetr
...
+	rms := metric.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		rm := rms.At(i)
+		namespace := GetNamespaceFromMetric(&rm)
+		ilms := rm.InstrumentationLibraryMetrics()
+		for j := 0; j < ilms.Len(); j++ {
+			ilm := ilms.At(j)
+			if ilm.IsNil() {
 				continue
 			}
```
Comment (kohrapha) [inline exporter/awsemfexporter/metric_translator.go:160]: I think you're missing the `Nil` check for `ilm.InstrumentationLibrary` here. Was there a reason you removed the logic for `instrumentationLibraryName`?

File: exporter/awsemfexporter/metric_translator.go [hunk 2]
```diff
@@ -140,48 +140,116 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, config *Config) ([]*CWMetr
...
+					case pdata.MetricDataTypeIntGauge:
+						dps = IntDataPointSlice{metric.IntGauge().DataPoints()}
+					case pdata.MetricDataTypeDoubleGauge:
+						dps = DoubleDataPointSlice{metric.DoubleGauge().DataPoints()}
+					case pdata.MetricDataTypeIntSum:
+						dps = IntDataPointSlice{metric.IntSum().DataPoints()}
+					case pdata.MetricDataTypeDoubleSum:
+						dps = DoubleDataPointSlice{metric.DoubleSum().DataPoints()}
+					case pdata.MetricDataTypeDoubleHistogram:
+						dps = DoubleHistogramDataPointSlice{metric.DoubleHistogram().DataPoints()}
```
Comment (kohrapha) [inline exporter/awsemfexporter/metric_translator.go:180]: You're missing a default case here.

File: exporter/awsemfexporter/metric_translator.go [hunk 3]
```diff
@@ -140,48 +140,116 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, config *Config) ([]*CWMetr
...
+			}
 		}
 	}
-	return cwMetricList, totalDroppedMetrics
 }
 
-func TranslateCWMetricToEMF(cwMetricLists []*CWMetrics) []*LogEvent {
-	// convert CWMetric into map format for compatible with PLE input
+// convert map of GroupedMetric objects into map format for compatible with PLE input
+func TranslateBatchedMetricToEMF(groupedMetricMap map[string]*GroupedMetric) []*LogEvent {
```
Comment (kohrapha) [inline exporter/awsemfexporter/metric_translator.go:213]: As we mentioned in the design doc, we don't need to translate GroupedMetrics to EMF.

**Review by kohrapha — CHANGES_REQUESTED** [review]
Thanks for cleaning up the PR. I think this PR still needs some work. Specifically, there's a few areas where you refactored the code but didn't copy over the changes correctly. I believe if you ran the tests, some tests should fail. I would suggest going through your CR and scanning to make sure you copied over code correctly, and running package tests to ensure all tests pass.

Also, I don't think we need the `TranslateBatchedMetricToEMF` function as we discussed in the design discussion that this PR is only concerned with _generating_ the `GroupedMetrics`.

File: exporter/awsemfexporter/metric_translator.go (hunk 2 above)
Comment (bryanlou) [reply exporter/awsemfexporter/metric_translator.go:180]: good catch

File: exporter/awsemfexporter/metric_translator.go (hunk 1 above)
Comment (bryanlou) [reply exporter/awsemfexporter/metric_translator.go:160]: added nil check

File: exporter/awsemfexporter/metric_translator.go [hunk 4]
```diff
@@ -140,48 +140,116 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, config *Config) ([]*CWMetr
...
 
-		metrics := ilm.Metrics()
-		for k := 0; k < metrics.Len(); k++ {
-			metric := metrics.At(k)
-			if metric.IsNil() {
-				totalDroppedMetrics++
+	var dps DataPoints
+	rms := metric.ResourceMetrics()
+	for i := 0; i < rms.Len(); i++ {
+		rm := rms.At(i)
```
[the remark this answers is not in the document]
Comment (bryanlou) [reply exporter/awsemfexporter/metric_translator.go:153]: added nil check

File: exporter/awsemfexporter/metric_translator.go [hunk 5]
```diff
@@ -140,48 +140,116 @@ func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, config *Config) ([]*CWMetr
 	if len(namespace) == 0 {
 		namespace = defaultNameSpace
 	}
+	return namespace
```
[the remark this answers is not in the document]
Comment (bryanlou) [reply exporter/awsemfexporter/metric_translator.go:143]: I think either works so going to leave it as it is.

File: exporter/awsemfexporter/metric_translator.go [hunk 6]
```diff
@@ -118,13 +123,8 @@ func (dps DoubleHistogramDataPointSlice) At(i int) DataPoint {
...
 
-// TranslateOtToCWMetric converts OT metrics to CloudWatch Metric format
-func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, config *Config) ([]*CWMetrics, int) {
-	var cwMetricList []*CWMetrics
-	totalDroppedMetrics := 0
-	namespace := config.Namespace
-	var instrumentationLibName string
-
+func GetNamespaceFromMetric(rm *pdata.ResourceMetrics) (string) {
+	var namespace string
```
[the remark this answers is not in the document]
Comment (bryanlou) [reply exporter/awsemfexporter/metric_translator.go:127]: changed to `validateNamespace`

File: exporter/awsemfexporter/metric_translator.go (hunk 3 above)
Comment (bryanlou) [reply exporter/awsemfexporter/metric_translator.go:213]: I'm going to leave this as unresolved, as the plan is to replace this after metric declaration and dimension rollup have been integrated into this

File: exporter/awsemfexporter/metric_translator.go [hunk 7]
```diff
@@ -55,15 +53,22 @@ type rateState struct {
...
+	Unit   string
+}
+
+// GroupedMetric defines
+type GroupedMetric struct {
+	Namespace    string
 	Timestamp    int64
-	Fields       map[string]interface{}
+	Labels      map[string]interface{}
+	Metrics     map[string]MetricInfo
```
[the remark this answers is not in the document]
Comment (bryanlou) [reply exporter/awsemfexporter/metric_translator.go:67]: Using pointer for GroupedMetric creation

File: exporter/awsemfexporter/emf_exporter.go [hunk 8]
```diff
@@ -172,26 +172,11 @@ func (emf *emfExporter) Start(ctx context.Context, host component.Host) error {
...
-		if len(cwm) > 0 && len(cwm[0].Measurements) > 0 {
-			namespace = cwm[0].Measurements[0].Namespace
-		}
-		// append all datapoint metrics in the request into CWMetric list
-		cwMetricLists = append(cwMetricLists, cwm...)
-	}
-
-	return TranslateCWMetricToEMF(cwMetricLists), totalDroppedMetrics, namespace
+	groupedMetricMap := make(map[string]*GroupedMetric)
+	totalDroppedMetrics = TranslateOtToGroupedMetric(metric, groupedMetricMap, config)
```
Comment (hdj630) [inline exporter/awsemfexporter/emf_exporter.go:177]: Since we are not going to make it work end2end, I think you can implement the `TranslateOtToGroupedMetric` only and leave the remaining code unchanged. I think that will be easier for raphael to pick and choose when integrate batching logic with metric selection.

File: exporter/awsemfexporter/emf_exporter.go [hunk 9]
```diff
@@ -172,26 +172,11 @@ func (emf *emfExporter) Start(ctx context.Context, host component.Host) error {
...
-		}
-		// append all datapoint metrics in the request into CWMetric list
-		cwMetricLists = append(cwMetricLists, cwm...)
-	}
-
-	return TranslateCWMetricToEMF(cwMetricLists), totalDroppedMetrics, namespace
+	namespace := config.Namespace
+	groupedMetricMap := make(map[string]*GroupedMetric)
+	groupedMetricMap, totalDroppedMetrics = TranslateOtToGroupedMetric(metric, config)
+	return TranslateBatchedMetricToEMF(groupedMetricMap), totalDroppedMetrics, namespace
```
Comment (kohrapha) [inline exporter/awsemfexporter/emf_exporter.go:179]: This still need to be reverted back right?

File: exporter/awsemfexporter/factory.go [hunk 10]
```diff
@@ -65,4 +65,4 @@ func createMetricsExporter(_ context.Context,
 	expCfg := config.(*Config)
 
 	return New(expCfg, params)
-}
+}
```
Comment (kohrapha) [inline exporter/awsemfexporter/factory.go:68]: Please undo this change.

File: exporter/awsemfexporter/metric_translator.go [hunk 11]
```diff
@@ -44,8 +45,6 @@ const (
 	// DimensionRollupOptions
 	ZeroAndSingleDimensionRollup = "ZeroAndSingleDimensionRollup"
 	SingleDimensionRollupOnly    = "SingleDimensionRollupOnly"
-
-	FakeMetricValue = 0
```
Comment (kohrapha) [inline exporter/awsemfexporter/metric_translator.go:48]: Why is this removed?

File: exporter/awsemfexporter/metric_translator.go [hunk 12]
```diff
@@ -118,13 +131,9 @@ func (dps DoubleHistogramDataPointSlice) At(i int) DataPoint {
...
 
-// TranslateOtToCWMetric converts OT metrics to CloudWatch Metric format
-func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, config *Config) ([]*CWMetrics, int) {
-	var cwMetricList []*CWMetrics
-	totalDroppedMetrics := 0
-	namespace := config.Namespace
-	var instrumentationLibName string
 
+// Validates namespace for given set of metrics from user config 
+func validateNamespace(rm *pdata.ResourceMetrics, namespace string) (string) {
```
Comment (kohrapha) [inline exporter/awsemfexporter/metric_translator.go:136]: Don't think we're _validating_ the namespace here. It looks to me like we're _retrieving_ it. So, something like `retrieveNamespace` or `parseNamespace` might be a better fit.

File: exporter/awsemfexporter/metric_translator.go [hunk 13]
```diff
@@ -421,3 +468,221 @@ func needsCalculateRate(pmd *pdata.Metric) bool {
...
+			string(pleMsg),
+		)
+		logEvent.LogGeneratedTime = time.Unix(0, metricCreationTime*int64(time.Millisecond))
+		ples = append(ples, logEvent)
+	}
+	return ples
+}
+
+// TranslateOtToCWMetric converts OT metrics to CloudWatch Metric format
+func TranslateOtToCWMetric(rm *pdata.ResourceMetrics, dimensionRollupOption string, namespace string) ([]*CWMetrics, int) {
```
Comment (kohrapha) [inline exporter/awsemfexporter/metric_translator.go:525]: Why is this inserted here? This might be because you removed it but then inserted it back in, causing the git diff to think this is new code. This makes it a bit difficult to review as we can't guarantee that this is the exact same code. Can you fix this?

File: exporter/awsemfexporter/metric_translator.go [hunk 14]
```diff
@@ -421,3 +468,221 @@ func needsCalculateRate(pmd *pdata.Metric) bool {
...
+				continue
+			}
+			cwMetrics := getCWMetrics(&metric, namespace, instrumentationLibName, dimensionRollupOption)
+			cwMetricList = append(cwMetricList, cwMetrics...)
+		}
+	}
+	return cwMetricList, totalDroppedMetrics
+}
+
+func TranslateCWMetricToEMF(cwMetricLists []*CWMetrics) []*LogEvent {
```
Comment (kohrapha) [inline exporter/awsemfexporter/metric_translator.go:572]: Likewise for this, can you replace this above so that git won't think that this is a new change?

File: exporter/awsemfexporter/metric_translator.go [hunk 15]
```diff
@@ -421,3 +468,221 @@ func needsCalculateRate(pmd *pdata.Metric) bool {
...
+			string(pleMsg),
+		)
+		logEvent.LogGeneratedTime = time.Unix(0, metricCreationTime*int64(time.Millisecond))
+		ples = append(ples, logEvent)
+	}
+	return ples
+}
+
+// Translates OTLP Metric to list of CW Metrics
+func getCWMetrics(metric *pdata.Metric, namespace string, instrumentationLibName string, dimensionRollupOption string) []*CWMetrics {
```
Comment (kohrapha) [inline exporter/awsemfexporter/metric_translator.go:599]: Same for this.

**Review by kohrapha — CHANGES_REQUESTED** [review]
It's a bit difficult to review as there's a lot of unnecessary code changes that might not be relevant to the PR.  Could you clean up this PR such that the diff we see is as easy to read as possible? For example, can you ensure that the 3 functions that you re-inserted are treated as the same code as before instead of being treated as new code?

**Comment by kohrapha** [comment]
Closing in favour of #95.
